File: cpp/filters/basic_filters.cpp
#include "filters/basic_filters.h"
#include <opencv2/imgproc.hpp>

// ---------------------------------------------------------------------------
// BrightnessFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> BrightnessFilter::defineSettings() const {
    return {
        { "brightness", "Brightness", SettingType::Number, 0.0, QVariant(), -255.0, 255.0, -255.0, 255.0, 1.0 }
    };
}

cv::Mat BrightnessFilter::process(const cv::Mat& image) {
    double b = get("brightness").toDouble();
    if (b == 0.0) return image.clone();

    cv::Mat img;
    image.convertTo(img, CV_32F);
    img += b;
    
    cv::Mat out;
    img.convertTo(out, CV_8U, 1.0, 0.0); // automatically clips 0-255
    return out;
}

// ---------------------------------------------------------------------------
// ContrastFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> ContrastFilter::defineSettings() const {
    return {
        { "contrast", "Contrast", SettingType::Number, 1.0, QVariant(), 0.0, 5.0, 0.0, 5.0, 0.1 }
    };
}

cv::Mat ContrastFilter::process(const cv::Mat& image) {
    double c = get("contrast").toDouble();
    if (std::abs(c - 1.0) < 1e-4) return image.clone();

    cv::Mat img;
    image.convertTo(img, CV_32F);
    img = (img - 127.5) * c + 127.5;

    cv::Mat out;
    img.convertTo(out, CV_8U);
    return out;
}

// ---------------------------------------------------------------------------
// InvertFilter
// ---------------------------------------------------------------------------
cv::Mat InvertFilter::process(const cv::Mat& image) {
    cv::Mat out;
    cv::bitwise_not(image, out);
    return out;
}

// ---------------------------------------------------------------------------
// ThresholdFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> ThresholdFilter::defineSettings() const {
    return {
        { "threshold", "Threshold", SettingType::Integer, 127, QVariant(), 0, 255, 0, 255, 1 }
    };
}

cv::Mat ThresholdFilter::process(const cv::Mat& image) {
    int t = get("threshold").toInt();
    cv::Mat out;
    cv::threshold(image, out, t, 255, cv::THRESH_BINARY);
    return out;
}

// ---------------------------------------------------------------------------
// UnsharpMaskFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> UnsharpMaskFilter::defineSettings() const {
    return {
        { "radius", "Radius", SettingType::Integer, 5, QVariant(), 1, 50, 1, 50, 1 },
        { "amount", "Amount", SettingType::Number,  1.5, QVariant(), 0.0, 5.0, 0.0, 5.0, 0.1 }
    };
}

cv::Mat UnsharpMaskFilter::process(const cv::Mat& image) {
    int r = get("radius").toInt();
    double a = get("amount").toDouble();

    int ksize = r * 2 + 1;
    cv::Mat blurred;
    cv::GaussianBlur(image, blurred, cv::Size(ksize, ksize), 0);

    cv::Mat imgFloat, blurredFloat;
    image.convertTo(imgFloat, CV_32F);
    blurred.convertTo(blurredFloat, CV_32F);

    cv::Mat sharpened = imgFloat + (imgFloat - blurredFloat) * a;
    
    cv::Mat out;
    sharpened.convertTo(out, CV_8U);
    return out;
}

File: python/app/filters/basic_filters.py
"""Basic Image Filters (Brightness, Contrast, Threshold, etc.)."""

import cv2
import numpy as np
from typing import List

from app.filters.base_filter import ImageFilter
from app.pfm import PFMSetting, SettingType


class BrightnessFilter(ImageFilter):
    @property
    def name(self) -> str: return "Brightness"

    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("brightness", "Brightness", SettingType.NUMBER, 0.0,
                       min_val=-255.0, max_val=255.0, step=1.0)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        b = self.get("brightness")
        if b == 0.0:
            return image
        
        if image.dtype != np.float32:
            img = image.astype(np.float32)
        else:
            img = image
            
        img = img + b
        return np.clip(img, 0, 255).astype(np.uint8)


class ContrastFilter(ImageFilter):
    @property
    def name(self) -> str: return "Contrast"

    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("contrast", "Contrast", SettingType.NUMBER, 1.0,
                       min_val=0.0, max_val=5.0, step=0.1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        c = self.get("contrast")
        if c == 1.0:
            return image
        
        if image.dtype != np.float32:
            img = image.astype(np.float32)
        else:
            img = image

        img = (img - 127.5) * c + 127.5
        return np.clip(img, 0, 255).astype(np.uint8)


class InvertFilter(ImageFilter):
    @property
    def name(self) -> str: return "Invert"

    @property
    def category(self) -> str: return "Color"

    def process(self, image: np.ndarray) -> np.ndarray:
        return 255 - image


class ThresholdFilter(ImageFilter):
    @property
    def name(self) -> str: return "Threshold"

    @property
    def category(self) -> str: return "Artistic"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("threshold", "Threshold", SettingType.INTEGER, 127,
                       min_val=0, max_val=255, step=1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        thresh = self.get("threshold")
        _, out = cv2.threshold(image, thresh, 255, cv2.THRESH_BINARY)
        return out


class UnsharpMaskFilter(ImageFilter):
    @property
    def name(self) -> str: return "Unsharp Mask"

    @property
    def category(self) -> str: return "Blur/Sharpen"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("radius", "Radius", SettingType.INTEGER, 5,
                       min_val=1, max_val=50, step=1),
            PFMSetting("amount", "Amount", SettingType.NUMBER, 1.5,
                       min_val=0.0, max_val=5.0, step=0.1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        r = self.get("radius")
        a = self.get("amount")
        
        # Kernel size must be odd
        ksize = r * 2 + 1
        
        blurred = cv2.GaussianBlur(image, (ksize, ksize), 0)
        img_float = image.astype(np.float32)
        blurred_float = blurred.astype(np.float32)
        
        sharpened = img_float + (img_float - blurred_float) * a
        return np.clip(sharpened, 0, 255).astype(np.uint8)
